test/michael_scott_queue/michael_scott_queue_dscheck.ml

```diff
@@ -1,151 +1,165 @@
+                let remaining = drain queue in
+                let popped = List.filter Option.is_some !popped in
+                List.length popped + remaining = items_total)))
+
+  let two_producers () =
+    Atomic.trace (fun () ->
+        let queue = Michael_scott_queue.create () in
+        let items_total = 4 in
+
+        (* producers *)
+        for _ = 1 to 2 do
+          Atomic.spawn (fun () ->
+              for _ = 1 to items_total / 2 do
+                Michael_scott_queue.push queue 0
+              done)
+        done;
+
+        (* checks*)
+        Atomic.final (fun () ->
+            Atomic.check (fun () ->
+                let remaining = drain queue in
+                remaining = items_total)))
+
+  let two_domains () =
+    Atomic.trace (fun () ->
+        let stack = Michael_scott_queue.create () in
+        let n1, n2 = (2, 1) in
+
+        (* two producers *)
+        let lists =
+          [
+            (List.init n1 (fun i -> i), ref []);
+            (List.init n2 (fun i -> i + n1), ref []);
+          ]
+        in
+        List.iter
+          (fun (lpush, lpop) ->
+            Atomic.spawn (fun () ->
+                List.iter
+                  (fun elt ->
+                    (* even nums belong to thr 1, odd nums to thr 2 *)
+                    Michael_scott_queue.push stack elt;
+                    lpop :=
+                      Option.get (Michael_scott_queue.pop_opt stack) :: !lpop)
+                  lpush)
+            |> ignore)
+          lists;
+
+        (* checks*)
+        Atomic.final (fun () ->
+            let lpop1 = !(List.nth lists 0 |> snd) in
+            let lpop2 = !(List.nth lists 1 |> snd) in
+
+            (* got the same number of items out as in *)
+            Atomic.check (fun () -> List.length lpop1 = n1);
+            Atomic.check (fun () -> List.length lpop2 = n2);
+
+            (* no element are missing *)
+            Atomic.check (fun () ->
+                let l1 = List.filter (fun i -> i < n1) lpop1 in
+                let l2 = List.filter (fun i -> i >= n1) lpop1 in
+                let l3 = List.filter (fun i -> i < n2) lpop2 in
+                let l4 = List.filter (fun i -> i >= n2) lpop2 in
+                let is_sorted l = List.sort (fun a b -> -compare a b) l = l in
+                is_sorted l1 && is_sorted l2 && is_sorted l3 && is_sorted l4)))
+
+  let tests name =
+    let open Alcotest in
     [
-      ( "basic",
+      ( "basic_" ^ name,
         [
           test_case "1-producer-1-consumer" `Slow producer_consumer;
           test_case "1-producer-1-consumer-peek" `Slow producer_consumer_peek;
           test_case "2-producers" `Slow two_producers;
           test_case "2-domains" `Slow two_domains;
         ] );
     ]
+end
+
+let () =
+  let module Safe = Dscheck_ms_queue (Michael_scott_queue) in
+  let safe_test = Safe.tests "safe" in
+  let module Unsafe = Dscheck_ms_queue (Michael_scott_queue_unsafe) in
+  let unsafe_test = Unsafe.tests "unsafe" in
```

Comment:
Not a major issue, but this would probably be a bit less verbose with first-class modules.  It could be like:

```ocaml
let safe_tests = tests "safe" (module Michael_scott_queue) in
let unsafe_tests = tests "unsafe" (module Michael_scott_queue_unsafe) in
...
```

The module language can be a bit cumbersome at times and in a case like this where we are only interested in a simple (non module) return value (a list of tests) first-class modules can be (IMHO) more convenient.